src/main/kotlin/org/jetbrains/research/testspark/appstarter/ProjectApplicationUtils.kt

```diff
@@ -0,0 +1,224 @@
+            ?: throw ProjectConfiguratorException("Can not open or import project from $projectPath.")
+        Disposer.register(parentDisposable) { closeProject(project) }
+
+        waitAllStartupActivitiesPassed(project)
+
+        ApplicationManager.getApplication().invokeAndWait {
+            VirtualFileManager.getInstance().refreshWithoutFileWatcher(false)
+        }
+
+        ApplicationManager.getApplication().invokeAndWait {
+            PatchProjectUtil.patchProject(project)
+        }
+
+        waitForInvokeLaterActivities()
+
+        return project
+    }
+
+    private suspend fun convertProject(projectPath: Path) {
+        val conversionService = ConversionService.getInstance()
+            ?: throw ProjectConfiguratorException("Can not convert project $projectPath")
+        val conversionResult = conversionService.convertSilently(projectPath, ConversionListenerImpl())
+        if (conversionResult.openingIsCanceled()) {
+            throw ProjectConfiguratorException("Project opening is canceled $projectPath")
+        }
+    }
+
+    private fun configureProjectEnvironment(projectPath: Path) {
+        for (configurator in CommandLineInspectionProjectConfigurator.EP_NAME.extensionList) {
+            val context = ConfiguratorContextImpl(projectPath)
+            if (configurator.isApplicable(context)) {
+                logger.info("Applying configurator ${configurator.name} to configure project environment $projectPath.")
+                configurator.configureEnvironment(context)
+            }
+        }
+    }
+
+    fun resolveProject(
+        project: Project,
+        configurator: CommandLineInspectionProjectConfigurator,
+        context: ConfiguratorContext
+    ) {
+        logger.info("Resolving project ${project.name}...")
+        logger.info("Applying configurator ${configurator.name} to resolve project ${project.name}...")
+        configurator.preConfigureProject(project, context)
+        configurator.configureProject(project, context)
+        waitForInvokeLaterActivities()
+        logger.info("Project ${project.name} was successfully resolved with configurator ${configurator.name}!")
+    }
+
+    private fun closeProject(project: Project) {
+        logger.info("Closing project $project...")
+        ApplicationManager.getApplication().assertIsNonDispatchThread()
+
+        ApplicationManager.getApplication().invokeAndWait {
+            ProjectManagerEx.getInstanceEx().forceCloseProject(project)
+        }
+    }
+
+    private suspend fun waitAllStartupActivitiesPassed(project: Project): Unit = suspendCoroutine {
+        logger.info("Waiting all startup activities passed $project...")
+        StartupManager.getInstance(project).runAfterOpened { it.resume(Unit) }
+        waitForInvokeLaterActivities()
+    }
+
+    /**
+     * Magic loop is used to wait all invoke later activities passed.
+     * No loop coses empty analyzer output as some project opening activities are not finished yet.
```

Comment:
I do not understand what is communicated in this sentence. Could you re-write it?